mautic/mautic
# Add Symfony Mailer Only

| Q                                      | A
| -------------------------------------- | ---
| Bug fix? (use the a.b branch)          | [ ]
| New feature/enhancement? (use the a.x branch)      | 5.x
| Deprecations?                          | Yes
| BC breaks? (use the c.x branch)        | BC breaks
| Automated tests included?              | Some automated tests are included, we need to add more to cover all the changes 
| Related user documentation PR URL      | https://github.com/mautic/user-documentation/pull/174
| Related developer documentation PR URL | [mautic/developer-documentation#... ](https://mautic-developer.readthedocs.io/en/5.x/components/emails.html#testing-email-transports)
| Issue(s) addressed                     | Fixes #... 



#### Description:

**Email Bundle**
In the `EmailBundle` we rely on `[Symfony/Mailer](https://symfony.com/doc/current/mailer.html)` and removed all the references to swiftmailer

**What has changed ?**
1. All the transports have been removed because we want them to be plugins outside the core of Mautic core
2. Mailer is integrated with messenger so emails are queued directly to the database (or your choice of queue).
3. Callbacks classes remained the same but their URLs will be changed depending on the transport name for example `amazon_api` has been changed to `ses+api`.
4. There is a single parameter that configures mailer in a format of DSN like this `ses+api:username:password@default?region=us-east-1`
5. **Support for File Spool and File queue has been dropped**, in order to restore the queue functionality PR #11598 should be merged and then queues will be used

**What is missing for this bundle**
1. Automated tests
2. Including other transports as plugins
3. Manual testing 
4. Plugins to include all the transports that Mautic has now

**This PR introduces the following improvements:**
1. Faster email sending, I tested with SES API and managed to get 3.5X speed increase compared to what we have in M4.
2. Less memory consumption as messenger compacts the emails when they are stored in the queue
3. Storage optimization, with Mautic 4 file spool, the disk may get filled with messages pending but with this change the messages will be loaded to Redis or the database and multiple processes can send at the same time.
4. We were able to run 32 processes to send emails simultaneously
6. This PR will be the door to introducing faster and more lightweight transports
7. We can use the transports that are maintained by Symfony without extra work on our side 


#### Steps to test this PR:


1. Open this PR on Gitpod or pull down for testing locally (see docs on testing PRs [here](https://contribute.mautic.org/contributing-to-mautic/tester))
2. Try installing Mautic from the UI and the CLI 
3. Send Emails from every possible place (config menu, campaign, segment, form, etc)
4. Make sure it works
5. Tips on the tests: include CC, BCC, attachments, emails with Unicode characters in the To, Reply-to, CC, From, BCC
6. **The email from should not include Unicode characters as specified in https://symfony.com/doc/5.4/mailer.html#email-addresses**




#### Immediate issues to fix after Symfony 5 upgrade
1. Make a better serialization of MauticMessage, right now it uses PHP native serialization, we need to convert it to JSON which is compact.
2. Include a bus for emails
3. Here is a sample of AWS SES transport and it shows how it inject itself to the refactored code https://github.com/mabumusa1/ses-plugin
